### PR TITLE
Meter import/export of cadence values and types

### PR DESCRIPTION
**Pull request by robert-e-davidson3**
Closes https://github.com/dapperlabs/cadence-private-issues/issues/38

## Description

Metering import and export of the cadence.Value and cadence.Type data structures.

Added expected weights to notion doc: https://www.notion.so/dapperlabs/3686f94810664b51833ba922ecde5552?v=bd3091ef17a04684ab96d4bc97d9d492
______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: values.go [hunk 1]
```diff
@@ -1286,10 +1636,23 @@ type Capability struct {
...
+	common.UseConstantMemory(gauge, common.MemoryKindCadenceCapability)
+	return NewUnmeteredCapability(path, address, borrowType)
+}
+
 func (Capability) isValue() {}
 
-func (Capability) Type() Type {
-	return CapabilityType{}
+func (v Capability) Type(gauge common.MemoryGauge) Type {
+	return NewCapabilityType(gauge, v.BorrowType) // TODO is this correct? original was nil for BorrowType
```
Comment (robert-e-davidson3) [inline values.go:1655]: I need advice here @turbolent.

File: encoding/json/decode.go [hunk 2]
```diff
@@ -333,12 +355,19 @@ func (d *Decoder) decodeInt64(valueJSON interface{}) cadence.Int64 {
 		panic(ErrInvalidJSONCadence)
 	}
 
-	return cadence.NewInt64(i)
+	return cadence.NewInt64(d.gauge, i)
 }
 
 func (d *Decoder) decodeInt128(valueJSON interface{}) cadence.Int128 {
 	bigInt := d.decodeBigInt(valueJSON)
```
Comment (turbolent) [inline encoding/json/decode.go:362]: The current code is a good start, in the future we actually also need to meter allocations here: The JSON string is converted to a bigint value, which already allocates, which should go into the function below:
https://github.com/onflow/cadence/pull/1611/files#diff-486c5071ade706b28d51beacaf8b7cbe4267a892e708877e4ef37e3cfbc464d4R367.

We can address this in a follow-up PR though, not a blocker for this one

**Review by turbolent — COMMENTED** [review]
It looks like this depends on the static type metering?

File: encoding/json/decode.go (hunk 2 above)
Comment (robert-e-davidson3) [reply encoding/json/decode.go]: I went ahead and did this change since I needed to fix up the PR anyway.

**Comment by robert-e-davidson3** [comment]
I redid the branch so it comes off of `feature/memory-metering`.

**Comment by turbolent** [comment]
This changes the API for the "public API", the `cadence` package, so labeling as a breaking change

**Review by turbolent — APPROVED** [review]
Nice work!

@SupunS could you please have a last look?

**Review by SupunS — COMMENTED** [review]
Nice! changes looks good!

I have one question/suggestion wrt API change: Since the value and type creation APIs are used across different toolings (sdk, etc.) and can impact many downstream codes, what if we keep the current constructors as is, and introduce a new 'metered' constructor.

e.g: for `UInt8`, keep `NewUInt8()` as is, and introduce a new constructor `NewMeteredUInt8(memoryGauge)` to be used internally within cadence. So the external users don't have any impact. This might be just a simple renaming, on top of this PR

**Comment by robert-e-davidson3** [comment]
> Nice! changes looks good!
> 
> I have one question/suggestion wrt API change: Since the value and type creation APIs are used across different toolings (sdk, etc.) and can impact many downstream codes, what if we keep the current constructors as is, and introduce a new 'metered' constructor.
> 
> e.g: for `UInt8`, keep `NewUInt8()` as is, and introduce a new constructor `NewMeteredUInt8(memorGauge)` to be used internally within cadence. So the external users don't have any impact. This might be just a simple renaming, on top of this PR

I like that idea.

File: values.go (hunk 1 above)
Comment (turbolent) [reply values.go:1655]: Yep, this makes sense 👍 It was previously not included

**Comment by turbolent** [comment]
@SupunS Great idea! That might reduce / avoid breakage 👍

File: values.go [hunk 3]
```diff
@@ -55,10 +56,15 @@ func NewVoid() Void {
...
+	common.UseConstantMemory(memoryGauge, common.MemoryKindCadenceVoid)
+	return NewVoid()
+}
+
 func (Void) isValue() {}
 
-func (Void) Type() Type {
-	return VoidType{}
+func (Void) Type(gauge common.MemoryGauge) Type {
+	return NewVoidType(gauge)
```
Comment (SupunS) [inline values.go:67]: Maybe we can introduce a new method here as well, keeping the old one as it is, similar to constructors.
Can do it in a subsequent PR.

**Review by SupunS — APPROVED** [review]

File: values.go (hunk 3 above)
Comment (robert-e-davidson3) [reply values.go:67]: Oh good point - that also alters the interface.

A second PR sounds good to me. This wouldn't be as easy a change as the constructor renamings.

Comment (robert-e-davidson3) [reply values.go:67]: https://github.com/dapperlabs/cadence-private-issues/issues/46

**Pull request merged**